File: AtCoder/ABC174/e.py
N, K = map(int, input().split())
A = list(map(int, input().split()))

def nibutan(f, ok, ng):
    while abs(ok - ng) > 1:
        mid = (ok + ng) // 2
        if f(mid):
            ok = mid
        else:
            ng = mid
    return ok

def check(target_length):
    count = 0
    for a in A:
        count += (a - 1) // target_length
    return count <= K

def main():
    ans = nibutan(check, max(A), 0)
    print(ans)

main()

File: AtCoder/ABC174/e.rb
N, K = gets.chomp.split.map(&:to_i)
A = gets.chomp.split.map(&:to_i)

def nibutan(ok, ng, &block)
  while (ok - ng).abs > 1
    mid = (ok + ng) / 2
    if yield(mid)
      ok = mid
    else
      ng = mid
    end
  end
  ok
end

def main
  ans = nibutan(A.max, 0) do |target_length|
    count = 0
    A.each do |a|
      count += (a - 1) / target_length
    end
    count <= K
  end
  puts ans
end

main()
